README.md
# SUIVEUR DE LIGNE

# Documentation

Le rapport est disponible [ici](./Rapport_suiveur_de_ligne_JorisIdjellidaine_LucianPetic.pdf)



docs/history.md
# History

## Week 1
1. Download OS from ```https://www.ev3dev.org/downloads/```
2. Need to create a bootable SD card.
3. To install the OS I follow those instructions ```http://docs.ev3dev.org/en/ev3dev-stretch/getting-started/ev3.html#installing-ev3dev```
4. I flashed the .iso to SD card, then I insert the SD card and installation started.
5. Unfortunately we can't connect to the device via SSH (no Wi-Fi), but we can connect via Bluetooth ```https://www.ev3dev.org/docs/tutorials/```
6. Via the Bluetooth connection we can get Internet, then connect via SSH
```shell script
$ ssh robot@ev3dev.local # password: maker
```

7. Programming language available: ```https://www.ev3dev.org/docs/programming-languages/```
8. To send a file:
```shell script
$ scp Hello.class robot@ev3dev.local:res # res is a folder
```
9. Finally, I launch a Python script to print ```Hello World!```
10. <p>Commit <a href="https://gaufre.informatique.univ-paris-diderot.fr/petic/idjellidaine-petic-plong-2020/commit/148f5e3927685c535ca3d141ec44e79ef2ff4c8f" rel="nofollow">148f5e39</a></p>

## Week 2

1. Need to launch at less one engine and get RGB from camera
2. I suggest taking a look on:  ```https://github.com/ev3dev/ev3dev-lang-python```.
3. I created a Makefile to:
    - **send** - send all files
    - **con** - connect via ssh
    - **id** - to copy your keys to the target server and avoid typing password

## Week 4

At the moment we can launch engines.

1. Get color from the sensor
2. Docs is here ```https://ev3dev-lang.readthedocs.io/projects/python-ev3dev/en/stable/sensors.html#ev3dev2.sensor.Sensor```
3. There are consts for color provides by EV3
4. I found a way to get closest tuple RGB from dict. Need to add another colors.


## Week 8

Added color learning and basic line following

1. Scan two colors from the sensor, the line and the floor
2. We can then control the motors and decide what to do based on the colors seen before and the color we see right now
3. Docs for the motors is here ```https://ev3dev-lang.readthedocs.io/projects/python-ev3dev/en/stable/motors.html```

## Week 12

Upgraded line following, now workd at a good pace

## Week 20

Upgraded to PID algorithm

Basic.py
#!/usr/bin/env python3

import time
from ev3dev2.motor import *
from ev3dev.ev3 import *
from time import sleep
from math import fabs


motor_left = LargeMotor(OUTPUT_A)
motor_right = LargeMotor(OUTPUT_D)
tank_drive = MoveSteering(OUTPUT_A, OUTPUT_D)
cs = ColorSensor()
cs.mode = 'RGB-RAW'
lcd = Screen()
seenColor = 0,0,0

SPEED = 20 #Maximum speed allowed to the robot in percentage (0 <= x <= 100)
MINSTEER = 10 #Minimum steer allowed to the robot in percentage (0 < x <= 100)
MAXSTEER = 65 #Maximum steer allowed to the robot in percentage (0 <= x <= 100)
STEERINGMULTIPLICATOR = 1 #Multiplicator for how fast the steering will be done

# Color Sensor need to be connected on INPUT 1 or 2
# Scans for 3 colors then returns them in the form a list of tuples (r,g,b) of size 3
# For proper operation, scan the floor first, then the line
def scanColors(rangeVal):
    colorsList = []
    
    # Put the color sensor into RGB mode.

    for _ in range(rangeVal):
        while True:
            if Button().any():
                break
            else:
                sleep(0.1)

        Sound.beep('-f 50')

        scanList = []
        for _ in range(20):
            scanList.append(getColor())
        colorsList.append(getAverageColor(scanList))
        print("Scanned",getAverageColor(scanList))
        
        Sound.beep('-f 50')

    return colorsList

# Takes a list of tuples (r,g,b) and returns an average tuple
def getAverageColor(colorsList):
    averageColor = [0, 0,0]

    for x in range(len(colorsList)):
        averageColor[0] += colorsList[x][0]
        averageColor[1] += colorsList[x][1]
        averageColor[2] += colorsList[x][2]

    return int(averageColor[0] / len(colorsList)), int(averageColor[1] / len(colorsList)), int(averageColor[2] / len(colorsList))

def getColor():
    return (cs.value(0), cs.value(1),cs.value(2))

# Returns True if colors are the same, else False
def isTheSameColor(color1, color2, margin):
    if color1[0] > (color2[0] + margin) or color1[0] < (color2[0] - margin):
        return False
    if color1[1] > (color2[1] + margin) or color1[1] < (color2[1] - margin):
        return False
    if color1[2] > (color2[2] + margin) or color1[2] < (color2[2] - margin):
        return False
    return True

#Find back the line and resume following
def helpImLost(colorsList,margin):
    tank_drive.on(0, SpeedPercent(0))
    sleep(0.5)
    tank_drive.on(100, SpeedPercent(10))

    while True:
        seenColor = getColor()
        if isTheSameColor(colorsList[1], seenColor, margin) or isTheSameColor(colorsList[2], seenColor, margin): #I found the line !
            break #When the line is found again, the robot isn't lost anymore and can resume line following

def getMargin(colorsList):
    margin = 80

    marginColorsList = colorsList.copy()
    del marginColorsList[3]
    marginColorsList.append(getAverageColor([colorsList[1],colorsList[2]]))
    marginColorsList.append(getAverageColor([colorsList[0],colorsList[2]]))

    while True:
        modified = False
        for x in range(len(marginColorsList)):
            for y in range(x+1,len(marginColorsList)):

                upperX = ((marginColorsList[x][0]+10),(marginColorsList[x][1]+10),(marginColorsList[x][2]+10))
                lowerX = ((marginColorsList[x][0]-10),(marginColorsList[x][1]-10),(marginColorsList[x][2]-10))
                upperY = ((marginColorsList[y][0]+10),(marginColorsList[y][1]+10),(marginColorsList[y][2]+10))
                lowerY = ((marginColorsList[y][0]-10),(marginColorsList[y][1]-10),(marginColorsList[y][2]-10))
                
                if(isTheSameColor(marginColorsList[x],marginColorsList[y],margin) or isTheSameColor(upperX,lowerY,margin) or isTheSameColor(lowerX,upperY,margin)):
                    margin -= 1
                    modified = True
                    #print(marginColorsList[x],",",x," is too close to ",marginColorsList[y],",",y,"with margin ",margin+1,"now reducing to ",margin)
                if(modified):break
            if(modified):break
        if(not modified):break
        
    print("Calculated margin :",max(10,margin - 3))
    return max(10,margin - 3)

# Improved followLine
def followLine(colorsList):

    steerAngle = MINSTEER #Angle of the steering
    STATE = -2  # -1 for negative steering, 0 for forward, 1 for positive steering
    lostCounter = 0 #Check if the robot is lost

    lastTimeOnFloor = time.time()
    previous_time = time.time()
    margin = getMargin(colorsList)

    timer = 0.05 #Given in seconds

    while True:

        #Time calculation
        now = time.time()
        timeChange = now - previous_time
        
        if(timeChange >= timer):
            seenColor = getColor() #Scan a color to then be used by the rest of the function

            if(isTheSameColor(colorsList[3], seenColor, margin)):
                seenColor = colorsList[1]

            #Lost timer
            if(not isTheSameColor(seenColor,colorsList[0],margin)):
                lastTimeOnFloor = time.time()

            #Lost launcher
            if(now - lastTimeOnFloor >= 1.5):
                helpImLost(colorsList,margin)


            #Cross detection, average
            if(margin >= 15 and (isTheSameColor(seenColor,getAverageColor([colorsList[1],colorsList[2]]),margin) or isTheSameColor(seenColor,getAverageColor([colorsList[0],colorsList[2]]),margin))):
                tank_drive.on_for_seconds(0,SpeedPercent(SPEED), 1, False)
                lastTimeOnFloor = time.time()

            #Cross detection, raw color
            elif(isTheSameColor(seenColor,colorsList[2],margin)):#Check for the orange line
                tank_drive.on_for_seconds(0,SpeedPercent(SPEED), 1, False)
                lastTimeOnFloor = time.time()


            if isTheSameColor(colorsList[1], seenColor, margin) or isTheSameColor(colorsList[3], seenColor, margin):
                if STATE != -1:
                    steerAngle = MINSTEER
                    STATE = -1
                    lostCounter = 0 #lostCounter is reinitialized upon seeing this color for the first time
                lostCounter += 1
                steerAngle = min(steerAngle+(lostCounter * STEERINGMULTIPLICATOR),MAXSTEER)
                tank_drive.on(steerAngle, SpeedPercent(SPEED))

            else : #If seenColor is the same color as the ground
                if STATE != 1: #Used to check for successive scans of the same color and to reinitialise upon scanning a new color
                    steerAngle = MINSTEER #steerAngle is reinitialized to MAXSTEER upon seeing this color for the first time
                    STATE = 1 #Each color is assigned to a different state
                    lostCounter = 0 #lostCounter is reinitialized upon seeing this color for the first time
                lostCounter += 1
                steerAngle = min(steerAngle+(lostCounter * STEERINGMULTIPLICATOR),MAXSTEER) #steerAngle is doubled for each successive scan of the same color but no greater than MAXSTEER
                tank_drive.on(-steerAngle, SpeedPercent(SPEED)) #Drive at an of steerAngle at a speed of speed

            

                

if __name__ == '__main__':

    Sound.beep("-f 200")
    Sound.beep("-f 300")
    Sound.beep("-f 400")

    colorsList = scanColors(4)

    while True:
        update = False
        if Button().enter:
            break
        if Button().up:
            SPEED += 5
            Sound.beep("-f "+str(SPEED * 10))
            update = True
        if Button().down:
            SPEED -= 5
            Sound.beep("-f "+str(SPEED * 10))
            update = True
        if Button().left:
            MAXSTEER -= 5
            Sound.beep("-f "+str(MAXSTEER * 10))
            update = True
        if Button().right:
            MAXSTEER += 5
            Sound.beep("-f "+str(MAXSTEER * 10))
            update = True

        if update:
            lcd.clear()
            lcd.draw.text((10,10),"Current speed = "+str(SPEED))
            lcd.draw.text((10,20),"Current steer = "+str(MAXSTEER))
            lcd.update()
        sleep(0.1)

    Sound.beep("-f 200")
    Sound.beep("-f 300")
    Sound.beep("-f 400")

    followLine(colorsList)

Scanner.py
#!/usr/bin/env python3

import time
from ev3dev2.motor import *
from ev3dev.ev3 import *
from time import sleep
from math import fabs

cs = ColorSensor()
cs.mode = 'RGB-RAW'

def scanColors(rangeVal):
    colorsList = []
    
    # Put the color sensor into RGB mode.

    for _ in range(rangeVal):
        while True:
            if Button().any():
                break
            else:
                sleep(0.1)

        scanList = []
        for _ in range(20):
            scanList.append(getColor())
        colorsList.append(getAverageColor(scanList))
        print("Scanned",getAverageColor(scanList))

    return colorsList
        
    

def getColor():
    return (cs.value(0), cs.value(1),cs.value(2))

def getAverageColor(colorsList):
    averageColor = [0, 0,0]

    for x in range(2):
        averageColor[0] += colorsList[x][0]
        averageColor[1] += colorsList[x][1]
        averageColor[2] += colorsList[x][2]

    return int(averageColor[0] / 2), int(averageColor[1] / 2), int(averageColor[2] / 2)

def isTheSameColor(color1, color2, margin):
    if color1[0] > (color2[0] + margin) or color1[0] < (color2[0] - margin):
        return False
    if color1[1] > (color2[1] + margin) or color1[1] < (color2[1] - margin):
        return False
    if color1[2] > (color2[2] + margin) or color1[2] < (color2[2] - margin):
        return False
    return True

def getMargin(colorsList):
    margin = 80

    while True:
        modified = False
        for x in range(len(colorsList)):
            for y in range(x+1,len(colorsList)):
                if(isTheSameColor(colorsList[x],colorsList[y],margin)):
                    margin -= 1
                    modified = True
                if(modified):break
            if(modified):break
        if(not modified):break
        print("Margin compute : ",margin)
    
    return margin - 5

if __name__ == '__main__':

    previous_time = time.time()

    print("Ready to scan")

    colorsList = scanColors(2)
    #colorsList.append((101, 111, 54))
    #colorsList.append((4, 4, 2))


    targetValue = getAverageColor([colorsList[0],colorsList[1]])
    print("current average is ",targetValue)
    previous_time = time.time()
    margin = getMargin(colorsList)
    print("current margin is",margin)

    #PID tuning

    timer = 0.05 #Given in seconds

    while True:
        #Time calculation
        now = time.time()
        timeChange = now - previous_time

        if(timeChange >= timer):
            seenColor = getColor()

            multiplicator = (100/abs((sum(list(targetValue))/3 - sum(colorsList[0])/3)) + abs(100/(sum(list(targetValue))/3 - sum(colorsList[1])/3)))/2

            inputColor = sum(list(seenColor))/3
            error = min(100,max(-100,(sum(list(targetValue))/3 - inputColor) * multiplicator))
            print("RGB :",seenColor,"Error :",error,', 1 :',isTheSameColor(seenColor,colorsList[1],margin),', 2 :',isTheSameColor(seenColor,colorsList[2],margin))


PID.py
#!/usr/bin/env python3

import time
from ev3dev2.motor import *
from ev3dev.ev3 import *
from time import sleep
from math import fabs


motor_left = LargeMotor(OUTPUT_A)
motor_right = LargeMotor(OUTPUT_D)
tank_drive = MoveSteering(OUTPUT_A, OUTPUT_D)
cs = ColorSensor()
cs.mode = 'RGB-RAW'
lcd = Screen()
seenColor = 0,0,0
margin = 0

Speed = 20

maxOutput = 100
minOutput = -maxOutput

isAverageOk = True

# Color Sensor need to be connected on INPUT 1 or 2
# Scans for 3 colors then returns them in the form a list of tuples (r,g,b) of size 3
# For proper operation, scan the floor first, then the line
def scanColors(rangeVal):
    colorsList = []
    
    # Put the color sensor into RGB mode.

    for _ in range(rangeVal):
        while True:
            if Button().any():
                break
            else:
                sleep(0.1)

        Sound.beep('-f 50')

        scanList = []
        for _ in range(20):
            scanList.append(getColor())
        colorsList.append(getAverageColor(scanList))
        #print("Scanned",getAverageColor(scanList))
        
        Sound.beep('-f 50')
    
    return colorsList

# Takes a list of tuples (r,g,b) and returns an average tuple
def getAverageColor(colorsList):
    averageColor = [0, 0,0]

    for x in range(2):
        averageColor[0] += colorsList[x][0]
        averageColor[1] += colorsList[x][1]
        averageColor[2] += colorsList[x][2]

    return int(averageColor[0] / 2), int(averageColor[1] / 2), int(averageColor[2] / 2)

def getColor():
    return (cs.value(0), cs.value(1),cs.value(2))

# Returns True if colors are the same, else False
def isTheSameColor(color1, color2, margin):
    if color1[0] > (color2[0] + margin) or color1[0] < (color2[0] - margin):
        return False
    if color1[1] > (color2[1] + margin) or color1[1] < (color2[1] - margin):
        return False
    if color1[2] > (color2[2] + margin) or color1[2] < (color2[2] - margin):
        return False
    return True

#Find back the line and resume following
def helpImLost(colorsList,margin):
    tank_drive.on(0, SpeedPercent(0))
    sleep(0.5)
    tank_drive.on(100, SpeedPercent(10))

    while True:
        seenColor = getColor()
        if isTheSameColor(colorsList[1], seenColor, margin) or isTheSameColor(getAverageColor([colorsList[0],colorsList[1]]), seenColor, margin) or isTheSameColor(colorsList[2], seenColor, margin): #I found the line !
            break #When the line is found again, the robot isn't lost anymore and can resume line following

def setTuning(kp, ki, kd, timer):
    return (kp,(ki*timer),(kd/timer))

def getMargin(colorsList):
    margin = 80
    isAverageOk = True

    marginColorsList = colorsList.copy()
    del marginColorsList[3]
    marginColorsList.append(getAverageColor([colorsList[1],colorsList[2]]))
    marginColorsList.append(getAverageColor([colorsList[0],colorsList[2]]))

    while True:
        modified = False
        for x in range(len(marginColorsList)):
            for y in range(x+1,len(marginColorsList)):

                upperX = ((marginColorsList[x][0]+10),(marginColorsList[x][1]+10),(marginColorsList[x][2]+10))
                lowerX = ((marginColorsList[x][0]-10),(marginColorsList[x][1]-10),(marginColorsList[x][2]-10))
                upperY = ((marginColorsList[y][0]+10),(marginColorsList[y][1]+10),(marginColorsList[y][2]+10))
                lowerY = ((marginColorsList[y][0]-10),(marginColorsList[y][1]-10),(marginColorsList[y][2]-10))
                
                if(isTheSameColor(marginColorsList[x],marginColorsList[y],margin) or isTheSameColor(upperX,lowerY,margin) or isTheSameColor(lowerX,upperY,margin)):
                    margin -= 1
                    modified = True
                    #print(marginColorsList[x],",",x," is too close to ",marginColorsList[y],",",y,"with margin ",margin+1,"now reducing to ",margin)
                if(modified):break
            if(modified):break
        if(not modified):break
        
    #print("Calculated margin :",margin)
    if(margin - 5 <= 12):
        isAverageOk = False
        margin = 80

        marginColorsList = colorsList.copy()
        del marginColorsList[3]

        while True:
            modified = False
            for x in range(len(marginColorsList)):
                for y in range(x+1,len(marginColorsList)):

                    upperX = ((marginColorsList[x][0]+10),(marginColorsList[x][1]+10),(marginColorsList[x][2]+10))
                    lowerX = ((marginColorsList[x][0]-10),(marginColorsList[x][1]-10),(marginColorsList[x][2]-10))
                    upperY = ((marginColorsList[y][0]+10),(marginColorsList[y][1]+10),(marginColorsList[y][2]+10))
                    lowerY = ((marginColorsList[y][0]-10),(marginColorsList[y][1]-10),(marginColorsList[y][2]-10))
                    
                    if(isTheSameColor(marginColorsList[x],marginColorsList[y],margin) or isTheSameColor(upperX,lowerY,margin) or isTheSameColor(lowerX,upperY,margin)):
                        margin -= 1
                        modified = True
                        #print(marginColorsList[x],",",x," is too close to ",marginColorsList[y],",",y,"with margin ",margin+1,"now reducing to ",margin)
                    if(modified):break
                if(modified):break
            if(not modified):break
    #print("Final calculated margin :",margin)
    return max(10,margin - 5),isAverageOk

def proportionalFollower(colorsList):

    targetValue = getAverageColor([colorsList[0],colorsList[1]])
    previous_time = time.time()
    
    
    #PID tuning
    timer = 0.05 #Given in seconds
    multiplicator = (100/abs((sum(list(targetValue))/3 - sum(colorsList[0])/3)) + abs(100/(sum(list(targetValue))/3 - sum(colorsList[1])/3)))/2
    settings = setTuning(0.45,0.9,0,timer)

    '''print("Margin :",margin)
    print("Target value :",targetValue)
    print("Multiplicator : (100/(",abs(sum(list(targetValue))/3)," - ",abs(sum(colorsList[0])/3),")) + (100/(",abs((sum(list(targetValue))))/3," - ",abs(sum(colorsList[1])/3),"))/2 = ",multiplicator)
    print("isAverageOk :",isAverageOk)'''

    Kp = settings[0]
    Ki = settings[1]
    Kd = settings[2]

    #Error manipulation
    error_sum = 0
    lastInput = sum(list(targetValue))/3

    #Lost count
    lastTimeOnFloor = time.time()
    lastSprint = time.time()

    while True:
        
        #Time calculation
        now = time.time()
        timeChange = now - previous_time
        
        if(timeChange >= timer):
            seenColor = getColor()
            trueColor = seenColor

            #Lost timer
            if(not isTheSameColor(trueColor,colorsList[0],22)):
                lastTimeOnFloor = time.time()

            #Lost launcher
            if(now - lastTimeOnFloor >= 1.5):
                helpImLost(colorsList,margin)
                error_sum = 0

            #Start line detection
            if(isTheSameColor(trueColor,colorsList[3],margin)):
                seenColor = colorsList[1]

            #Cross detection, last ressort, only used when the margin of error is too low and the gap has to crossed, cannot be used at high speeds
            elif((not isAverageOk or margin <= 12) and (isTheSameColor(trueColor,colorsList[1],margin))):
                if(now - lastSprint >= 2):
                    tank_drive.on_for_seconds(0,SpeedPercent(Speed), 1, False)
                    lastSprint = time.time()
                lastTimeOnFloor = time.time()

            #Cross detection, average, cannot be used when the margin of error is too low
            elif(isAverageOk and (isTheSameColor(trueColor,getAverageColor([colorsList[1],colorsList[2]]),margin) or isTheSameColor(trueColor,getAverageColor([colorsList[0],colorsList[2]]),margin))):
                if(now - lastSprint >= 2):
                    tank_drive.on_for_seconds(0,SpeedPercent(Speed), 1, False)
                    lastSprint = time.time()
                lastTimeOnFloor = time.time()

            #Cross detection, raw color
            elif(isTheSameColor(trueColor,colorsList[2],margin)):
                if(now - lastSprint >= 2):
                    tank_drive.on_for_seconds(0,SpeedPercent(Speed), 1, False)
                    lastSprint = time.time()
                lastTimeOnFloor = time.time()

            #Error calculations
            inputColor = sum(list(seenColor))/3
            error = min(100,max(-100,(sum(list(targetValue))/3 - inputColor) * multiplicator))
            error_sum += error * Ki #Ki decreases rampup time
            error_sum = min(max(error_sum,-65),65)#Max/min value for Ki error                
            derivative_input = inputColor - lastInput

            #PID calculations
            output = Kp * error + error_sum + Kd * derivative_input
            output = min(max(output,minOutput),maxOutput)

            #print(seenColor," | ",(int)(Kp * error)," + ",(int)(error_sum)," + ",(int)(Kd * derivative_input)," = ",(int)(output))
            #print(time.time() - startingTime,",",u,",",error)
            #print(Kp * error,"|",(int)(error_sum),"|",(int)(Kd * derivative_input),"|",(int)(output)," | ",speed)

            #Direction input
            tank_drive.on(output, SpeedPercent(Speed))

            #End of loop operations
            error_sum = error_sum * 0.925
            previous_time = now
            lastInput = inputColor

if __name__ == '__main__':

    Sound.beep("-f 200")
    Sound.beep("-f 300")
    Sound.beep("-f 400")

    colorsList = scanColors(4)
    margin, isAverageOk = getMargin(colorsList)

    lcd.clear()
    lcd.draw.text((10,10),"Current speed = "+str(Speed))
    lcd.draw.text((10,20),"Current margin = "+str(margin))
    lcd.draw.text((10,30),"IsAverageOk = "+str(isAverageOk))
    lcd.update()

    while True:
        update = False
        if Button().enter:
            break
        if Button().up:
            Speed += 5
            Sound.beep("-f "+str(Speed * 10))
            update = True
        if Button().down:
            Speed -= 5
            Sound.beep("-f "+str(Speed * 10))
            update = True
        if Button().left:
            margin -= 1
            Sound.beep("-f "+str(margin * 10))
            update = True
        if Button().right:
            margin += 2
            Sound.beep("-f "+str(margin * 10))
            update = True

        if update:
            lcd.clear()
            lcd.draw.text((10,10),"Current speed = "+str(Speed))
            lcd.draw.text((10,20),"Current margin = "+str(margin))
            lcd.draw.text((10,30),"IsAverageOk = "+str(isAverageOk))
            lcd.update()
        sleep(0.25)

    Sound.beep("-f 200")
    Sound.beep("-f 300")
    Sound.beep("-f 400")

    proportionalFollower(colorsList)